### ISSUE TITLE
no convenient stratified sampling ways

### ISSUE THREAD

**Issue by ayue2019** [mlr-org/mlr3book#198]
If I just want to split the full data into a training_set and a testing_set using stratified sampling, I found it complicated. I have to do like this:

`task <- TaskClassif$new(id = "wine_data", backend = wine, target = "class")

task$col_roles$stratum <- task$target_names
rs <- rsmp("holdout")
rs$instantiate(task)

train_set <- rs$train_set(1)
test_set <- rs$test_set(1)

prop.table(table(task$truth()))
prop.table(table(task$truth(train_set)))
prop.table(table(task$truth(test_set)))
`
it's also not convenient if I want to do this:
split a data into training data and test data, 
then train a model using 10-folds cross-validation based on training data(not full data),
then test the model performance using test data.

**Comment by pfistfl** [comment]
The way you solve the problem is also slightly against the `mlr3` design philosophy I would conjecture.

The following takes care of running `train` and `test` for you!

```r
task = tsk("wine")
task$col_roles$stratum <- task$target_names
rs <- rsmp("holdout")
resample(task, lrn("classif.rpart"), rs)
```

The 2nd thing does not really make sense, as the `10-fold CV` already yields an unbiased estimator of model performance, there is no need for an additional test split. This would only be required if you do model selection or tuning with your CV10. In that case, I would suggest to use an `AutoTuner` with `mlr3pipelines`. 

> then train a model using 10-folds cross-validation based on training data(not full data),
> then test the model performance using test data. 

Which of the 10 models would you like to test on? all? CV10 gives you 10 models.
If you are interested in the performance of each model trained, during CV on a single test-set, you can do the following:
```r
rs$instantiate(task)
# Do not use test_set during training.
task$set_row_role(rs$test_set(1), "validation")
# Get 10 models via 10-fold CV, store all
res = resample(task, lrn("classif.rpart"), rsmp("cv"), store_models=TRUE)
# Predict using test set.
sapply(res$learners, function(x) x$predict(task, row_id = rs$test_set(1)))
```

**Comment by ayue2019** [comment]
Thanks very much for your considerate answer.
I think I haven't been familiar with mlr3 package and its design philosophy enough. Your answer really helps a lot.


As for the second question, it' been my bad habit to do that when I train a model. I think it's time to correct myself.


It's exciting to have mlr3 package for machine learning in R. I hope it will be better and better.








------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "mlr-org/mlr3book"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月21日(星期三) 晚上8:17
收件人:&nbsp;"mlr-org/mlr3book"<mlr3book@noreply.github.com&gt;;
抄送:&nbsp;"Li Yue"<1272246204@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [mlr-org/mlr3book] no convenient stratified sampling ways (#198)





 
The way you solve the problem is also slightly against the mlr3 design philosophy I would conjecture.
 
The following takes care of running train and test for you!
 task = tsk("wine") task$col_roles$stratum <- task$target_names rs <- rsmp("holdout") resample(task, lrn("classif.rpart"), rs)
 
The 2nd thing does not really make sense, as the 10-fold CV already yields an unbiased estimator of model performance, there is no need for an additional test split. This would only be required if you do model selection or tuning with your CV10. In that case, I would suggest to use an AutoTuner with mlr3pipelines.
  
then train a model using 10-folds cross-validation based on training data(not full data),
 then test the model performance using test data.
  
Which of the 10 models would you like to test on? all? CV10 gives you 10 models.
 If you are interested in the performance of each model trained, during CV on a single test-set, you can do the following:
 rs$instantiate(task) # Do not use test_set during training. task$set_row_role(rs$test_set(1), "validation") # Get 10 models via 10-fold CV, store all res = resample(task, lrn("classif.rpart"), rsmp("cv"), store_models=TRUE) # Predict using test set. sapply(res$learners, function(x) x$predict(task, row_id = rs$test_set(1)))

**Comment by pfistfl** [comment]
You are welcome. I will close this here for now!
